frontend/model/contracts/group.js

```diff
@@ -312,32 +312,138 @@ sbp('chelonia/defineContract', {
-
-        return periodStampGivenDate({
-          recentDate,
-          periodStart: distributionDate,
-          periodLength: distributionPeriodLength
-        })
+        if (!distributionDate) return
+        const sortedPeriodKeys = getters.groupSortedPeriodKeys
+        // Maybe the distribution date has just been updated,
+        // but the corresponding period has not started or is not stored yet.
+        // So we try to use the distribution date first.
+        if (recentDate >= distributionDate) {
+          // Only extrapolate one period length in the future.
+          const extrapolatedDistributionDate = addTimeToDate(
+            distributionDate, distributionPeriodLength
+          ).toISOString()
+          if (recentDate >= extrapolatedDistributionDate) {
+            return dateToPeriodStamp(extrapolatedDistributionDate)
+          }
+          return dateToPeriodStamp(distributionDate)
+        }
+        // For dates before the distribution date, we check the stored data first,
+        // as simply substracting the default period length won't give the previous period stamp
+        // if the distribution date was updated during that period.
+        if (!sortedPeriodKeys.length) {
+          // Looks like we're in the waiting period but haven't stored it yet.
+          const waitingPeriodStamp = dateToPeriodStamp(
+            addTimeToDate(distributionDate, -distributionPeriodLength)
+          )
+          return recentDate >= dateFromPeriodStamp(waitingPeriodStamp).toISOString() ? waitingPeriodStamp : undefined
+        }
+        const oldestKnownStamp = sortedPeriodKeys[0]
+        const latestKnownStamp = sortedPeriodKeys[sortedPeriodKeys.length - 1]
+        if (recentDate < oldestKnownStamp) return
+        if (recentDate >= latestKnownStamp) {
+          const extrapolatedPeriodStamp = dateToPeriodStamp(
+            addTimeToDate(dateFromPeriodStamp(latestKnownStamp), distributionPeriodLength)
+          )
+          if (recentDate >= extrapolatedPeriodStamp) return extrapolatedPeriodStamp
+          return latestKnownStamp
+        }
+        for (let i = 1; i < sortedPeriodKeys.length; i++) {
+          if (recentDate < sortedPeriodKeys) return sortedPeriodKeys[i - 1]
+        }
+        // This should not happen
       }
     },
+    // May return 'undefined', in which case the caller should check archived data.
     periodBeforePeriod (state, getters) {
       return (periodStamp: string) => {
-        const len = getters.groupSettings.distributionPeriodLength
-        return dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(periodStamp), -len))
+        if (!isPeriodStamp(periodStamp)) throw new TypeError('must be periodStamp')
+        const { distributionDate, distributionPeriodLength } = getters.groupSettings
+        if (!distributionDate) return
+        // This is not always the current period stamp.
+        const distributionDateStamp = dateToPeriodStamp(distributionDate)
```

Comment:
I don't think you need to do this conversion, because `distributionDate` is not a `Date` object. It is in fact a period stamp that you can use directly.

Reply:
I thought `distributionDate` was an ISO date string, which happens to also be a period stamp but is conceptually a different thing (as you explained to me). It made sense to me since `distributionDate` comes from the group settings, whereas "period stamp" is a more internal type I think.

So anywhere I had the chance to compare a `distributionDate` value against a date directly, I didn't do any conversion.
I will likely now have to add a few of such conversions there if I start to consider `distributionDate` as being of the "period stamp" type.

Reply:
@snowteamer I think we made a mistake in how we named that variable... it is in fact a period stamp... and it happens to be a date represented as an ISO string.

Maybe you can add a comment to the constructor to that effect? (Renaming it now is not a good idea since there are live groups running).

Also, it's a bit confusing with how we've implemented `dateToPeriodStamp`, which takes either a `Date` or a `string`:

```js
export function dateToPeriodStamp (date: string | Date): string {
  return new Date(date).toISOString()
}
```

Reply:
Ok, thanks for explaining! 😄